### DISCUSSION TITLE
Move from #include magic to templates?

### DISCUSSION THREAD

**Discussion by mreineck** [flatironinstitute/finufft#444] [Ideas]
I see that templates are being used more extensively in `finufft` development recently, and based on that I'd like to suggest to use templating also for the implementation of single- and double-precision NUFFT routines. Up to now, this is done via compiling some source files twice, with different macro definitions for `FLT`, `CPX`, etc. Switching to templates would most likely simplify the build machinery quite a bit, probably testing as well.
It would also open up a way to differentiate between the accuracy of the strength and position arrays, allowing, e.g., for 1D NUFFTS with single precision complex data, but double-precision locations of nonuniform points, which can be extremely handy in many circumstances. But that would of course (if at all) be a second step.
Switching to templates may look problematic when considering the `finufft` C bindings and the interface to FFTW, but this can be done witout much effort and I'd be happy to help with this.

**Comment by DiamonDinoia** [comment]
Yes, that is a good idea. I am already moving in that direction. At some point, we will switch completely.
